python/lsst/daf/ingest/ingestCatalogTask.py

```diff
@@ -0,0 +1,580 @@
+    if sz > 65535:
+        # If necessary, longer strings could be ingested as TEXT.
+        raise RuntimeError("String field is too large for ingestion")
+    elif sz == 0:
+        raise RuntimeError("String field has zero size")
+    # A string containing trailing spaces cannot round-trip to a CHAR
+    # column and back. Therefore, use VARCHAR. Also, assume strings are
+    # ASCII for now.
+    return "VARCHAR({}) CHARACTER SET ascii COLLATE ascii_bin NOT NULL".format(sz)
+
+def _sqlTypeForArray(formatChar, field):
+    """Helper function for computing the SQL column type of an array valued field.
+    See https://docs.python.org/2/library/struct.html#format-characters for possible
+    format characters. The formatChar passed in must correspond to the array element
+    type.
+    """
+    sz = field.getSize()
+    if sz == 0:
+        return "BLOB NOT NULL"
+    sz *= struct.calcsize("<" + formatChar)
+    if sz > 65535:
+        raise RuntimeError("Array field is too large for ingestion")
+    return "BINARY({}) NOT NULL".format(sz)
+
+
+"""A dictionary mapping afw field type strings to corresponding formatters.
+
+This dictionary is used by IngestCatalogTask to determine how to format each
+type of field in the input catalog.  If new field types are added to the afw
+table library, they should also be added here.
+"""
+fieldFormatters = dict(
+    U = FieldFormatter(lambda f: "SMALLINT UNSIGNED NOT NULL", lambda v: str(v)),
+    I = FieldFormatter(lambda f: "INT NOT NULL", lambda v: str(v)),
+    L = FieldFormatter(lambda f: "BIGINT NOT NULL", lambda v: str(v)),
+    F = FieldFormatter(lambda f: "FLOAT", lambda v: _formatNumber("{:.9g}", v)),
+    D = FieldFormatter(lambda f: "DOUBLE", lambda v: _formatNumber("{:.17g}", v)),
+    Flag = FieldFormatter(lambda f: "BIT NOT NULL", lambda v: "1" if v else "0"),
+    Angle = FieldFormatter(lambda f: "DOUBLE", lambda v: _formatNumber("{:.17g}", v.asDegrees())),
+    String = FieldFormatter(_sqlTypeForString, _formatString),
+    ArrayU = FieldFormatter(lambda f: _sqlTypeForArray("H", f), lambda v: _formatArray("H", v)),
+    ArrayI = FieldFormatter(lambda f: _sqlTypeForArray("i", f), lambda v: _formatArray("i", v)),
+    ArrayF = FieldFormatter(lambda f: _sqlTypeForArray("f", f), lambda v: _formatArray("f", v)),
+    ArrayD = FieldFormatter(lambda f: _sqlTypeForArray("d", f), lambda v: _formatArray("d", v)),
+)
+
+
+def canonicalizeFieldName(fieldName):
+    """Return a MySQL-compatible version of the schema field name, for now by
+    changing any non-word characters to underscores.
+    """
+    return re.sub(r"[^\w]", "_", fieldName)
+
+def quoteIdentifier(identifier):
+    """Return a MySQL compatible version of the given identifier by quoting
+    it with back-ticks (and escaping embedded back-ticks).
+    """
+    return "`" + identifier.replace("`", "``") + "`"
+
+def aliasesFor(name, mappings):
+    """Compute and return the set of possible aliases for the given field name.
```

Comment:
So, I've fixed this particular alias cycle on [DM-3400](https://jira.lsstcorp.org/browse/DM-3400), which I'd put out for review right now if I wasn't trying to figure out a way to deal with the larger picture too on [DM-3401](https://jira.lsstcorp.org/browse/DM-3401), and that might impact the other issue.

The quick summary is that this kind of cycle was not supposed to happen, and there was actually already code that was supposed to prevent it, but it was broken.  And while we do want simple aliases for more than just backwards compatibility, the only use case we have right now for _recursive_ alias expansion is backwards compatibility.  With that in mind, it might be appropriate for the ingest code to simply assume alias expansion is not recursive, but I hope to have a more definitive answer on that subject by the end of the day.